src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java

```diff
@@ -233,13 +247,70 @@ private void cancel() {
 
     }
 
+    private class Callback2 extends BodyExecutionCallback {
+
+        @Override
+        public void onSuccess(StepContext context, Object result) {
+            if (killer != null) {
+                killer.cancel(true);
+                killer = null;
+            }
+            context.onSuccess(result);
+        }
+
+        @Override
+        public void onFailure(StepContext context, Throwable t) {
+            if (killer != null) {
+                killer.cancel(true);
+                killer = null;
+            }
+            if (t instanceof FlowInterruptedException) {
+                /*
+                 * Check whether the FlowInterruptedException has propagated past the corresponding
+                 * timeout step.
+                 */
+                FlowNode flowNode = null;
+                try {
+                    flowNode = getContext().get(FlowNode.class);
+                } catch (IOException | InterruptedException e) {
+                    LOGGER.log(Level.WARNING, null, e);
+                }
+                if (flowNode != null) {
+                    String nodeId = flowNode.getId();
+                    for (CauseOfInterruption cause : ((FlowInterruptedException) t).getCauses()) {
+                        if (cause instanceof ExceededTimeout) {
+                            ExceededTimeout exceededTimeout = (ExceededTimeout) cause;
+                            if (nodeId.equals(exceededTimeout.getNodeId())) {
+                                ((FlowInterruptedException) t).setActualInterruption(false);
+                            }
+                        }
+                    }
+                }
+            }
+            context.onFailure(t);
+        }
+
+        private static final long serialVersionUID = 1L;
+    }
+
     /**
      * Common cause in this step.
      */
     public static final class ExceededTimeout extends CauseOfInterruption {
 
         private static final long serialVersionUID = 1L;
 
+        /** Flow node ID of the corresponding timeout block */
+        private String nodeId;
+
+        private String getNodeId() {
+            return nodeId;
+        }
+
+        private void setNodeId(String nodeId) {
```

Comment:
Would make code simpler to just move this into the constructor (and make the field `final`).